File: LeetCode/39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
public:

    vector<vector<int>> combinationSum(vector<int>& v, int tar) {

        int sum=0;
        vector<int> temp;
        vector<vector<int>> ans;

        rec(v, tar, sum, temp, ans, 0);

        return ans;

     

        
    }


    void rec(vector<int>& v, int tar, int sum, vector<int> temp, vector<vector<int>> &ans, int i)
    {

        if(sum>tar)
        {
            return;
        }

        if(sum==tar)
        {
            ans.push_back(temp);
            return;
        }

        for(int j=i; j<v.size(); j++)
        {
            temp.push_back(v[j]);
            rec(v, tar, sum+v[j], temp, ans, j);
            temp.pop_back();
        }

        
      
    }






};

// @lc code=end

File: LeetCode/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start
class Solution:
    def combinationSum(self, arr: List[int], x: int) -> List[List[int]]:

        ans =[]


        def rec(arr, path, ind, x ):
            if x < 0:
                return
            if x == 0:
                # ans.append(path[:])
                ans.append(path)
                return
            for i in range(ind, len(arr)):
                rec(arr, path + [arr[i]], i, x - arr[i])  
               
               
               
                # path.append(arr[i])

                # rec(arr, path, i, x - arr[i])
                # path.pop()




        rec(arr, [], 0, x )

        return ans











        
# @lc code=end

